File: roc_app/GL/GLRenderbuffer.cpp
#include "stdafx.h"

#include "GLRenderbuffer.h"

GLuint GLRenderbuffer::ms_activeName = 0U;

GLRenderbuffer::GLRenderbuffer()
{
    m_name = 0U;
}

bool GLRenderbuffer::Create(GLenum p_format, GLsizei p_width, GLsizei p_height)
{
    if(m_name == 0U)
    {
        glGenRenderbuffers(1, &m_name);
        if(m_name != 0U)
        {
            glBindRenderbuffer(GL_RENDERBUFFER, m_name);
            glRenderbufferStorage(GL_RENDERBUFFER, p_format, p_width, p_height);

            glBindRenderbuffer(GL_RENDERBUFFER, ms_activeName);
        }
    }

    return (m_name != 0U);
}

bool GLRenderbuffer::Destroy()
{
    if(m_name != 0U)
    {
        if(ms_activeName == m_name)
        {
            ms_activeName = 0U;
            glBindRenderbuffer(GL_RENDERBUFFER, ms_activeName);
        }

        glDeleteRenderbuffers(1, &m_name);
        m_name = 0U;
    }

    return (m_name == 0U);
}

GLuint GLRenderbuffer::GetName() const
{
    return m_name;
}

bool GLRenderbuffer::Bind()
{
    if((m_name != 0U) && (ms_activeName != m_name))
    {
        glBindRenderbuffer(GL_RENDERBUFFER, m_name);
        ms_activeName = m_name;
    }

    return (m_name != 0U);
}

void GLRenderbuffer::Reset()
{
    if(ms_activeName != 0U)
    {
        ms_activeName = 0U;
        glBindRenderbuffer(GL_RENDERBUFFER, ms_activeName);
    }
}

File: roc_app/Engine/OGL/GLRenderBuffer.cs
using OpenGL;
using System;

namespace ROC.Engine.OGL
{
    internal sealed class GLRenderBuffer
    {
        [ThreadStatic]
        static uint ms_activeID = 0U;

        uint m_internalID = 0U;

        public bool IsValid => (m_internalID != 0U);
        internal uint InternalD => m_internalID;

        public GLRenderBuffer()
        {
        }

        public void Create(InternalFormat p_format, int p_width, int p_height)
        {
            if(m_internalID != 0U)
                return;

            m_internalID = Gl.GenRenderbuffer();
            if(m_internalID == 0U)
                return;

            Gl.BindRenderbuffer(RenderbufferTarget.Renderbuffer, m_internalID);
            Gl.RenderbufferStorage(RenderbufferTarget.Renderbuffer, p_format, p_width, p_height);
            Gl.BindRenderbuffer(RenderbufferTarget.Renderbuffer, ms_activeID);
        }

        public void Destroy()
        {
            if(m_internalID == 0U)
                return;

            Gl.DeleteRenderbuffers(m_internalID);

            if(ms_activeID == m_internalID)
            {
                Gl.BindRenderbuffer(RenderbufferTarget.Renderbuffer, 0U);
                m_internalID = 0U;
            }
        }

        public void Bind()
        {
            if(m_internalID == 0U)
                return;

            if(ms_activeID == m_internalID)
                return;

            Gl.BindRenderbuffer(RenderbufferTarget.Renderbuffer, m_internalID);
            ms_activeID = m_internalID;
        }

        public static void Reset()
        {
            if(ms_activeID == 0U)
                return;

            Gl.BindRenderbuffer(RenderbufferTarget.Renderbuffer, 0U);
            ms_activeID = 0U;
        }
    }
}
